src/OpenTelemetry.Instrumentation.Wcf/Implementation/RequestTelemetryStateTracker.cs

```diff
@@ -0,0 +1,171 @@
+                    {
+                        timedOutEntries = new List<Tuple<EntryTimeoutProperties, Entry>>();
+                    }
+
+                    timedOutEntries.Add(new(entryTimeoutProps, OutstandingRequestStates[entryTimeoutProps.MessageId]));
+                }
+                else
+                {
+                    nextToExpire = entryTimeoutProps;
+                    break;
+                }
+            }
+
+            foreach (var entry in timedOutEntries ?? Enumerable.Empty<Tuple<EntryTimeoutProperties, Entry>>())
+            {
+                OutstandingRequestStates.Remove(entry.Item1.MessageId);
+                TimeoutQueue.Remove(entry.Item1);
+            }
+
+            // when there's no more outstanding requests we set time timer to infinite, effectively disabling it until another request arrives
+            var nextExpiry = nextToExpire?.ExpiresAt ?? Timeout.Infinite;
+            SetTimer(nextExpiry);
+        }
+
+        timedOutEntries?.ForEach(entry => entry.Item1.TimeoutCallback(entry.Item2.State));
+    }
+
+    private static void SetTimerEarlierIfNeeded(long dueAt)
+    {
+        if (currentTimerDueAt == Timeout.Infinite || currentTimerDueAt > dueAt)
+        {
+            SetTimer(dueAt);
+        }
+    }
+
+    private static void SetTimer(long dueAt)
+    {
+        long dueIn;
+        if (dueAt == Timeout.Infinite)
+        {
+            dueIn = Timeout.Infinite;
+        }
+        else
+        {
+            var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+            dueIn = Math.Max(0, dueAt - now) + 50; // add 50ms here because the timer is not precise and often fires just a bit early
+        }
+
+        Timer.Change(dueIn, Timeout.Infinite);
+        currentTimerDueAt = dueAt;
+    }
+
+    private class Entry
+    {
+        public readonly RequestTelemetryState State;
+        public readonly EntryTimeoutProperties TimeoutProperties;
+
+        public Entry(RequestTelemetryState state, EntryTimeoutProperties timeoutProperties)
+        {
+            this.State = state;
+            this.TimeoutProperties = timeoutProperties;
+        }
+    }
+
+    private class EntryTimeoutProperties : IComparable
+    {
+        public readonly string MessageId;
+        public readonly long ExpiresAt;
+        public readonly Action<RequestTelemetryState> TimeoutCallback;
+
+        public EntryTimeoutProperties(string messageId, long expiresAt, Action<RequestTelemetryState> timeoutCallback)
+        {
+            this.MessageId = messageId;
+            this.ExpiresAt = expiresAt;
+            this.TimeoutCallback = timeoutCallback;
+        }
+
+        public int CompareTo(object obj)
+        {
+            var other = (EntryTimeoutProperties)obj;
+            var result = this.ExpiresAt.CompareTo(other.ExpiresAt);
+            if (result == 0)
+            {
+                result = string.Compare(this.MessageId, other.MessageId, StringComparison.Ordinal);
```

Comment:
This is unnecessary. A comparison result of 0 means they exist "in the same position" which in practive means the order they would appear in the SortedSet would be arbitrary. You don't care what order two items with the same expiry time will appear, so skip this extra calculation and return 0.

Reply:
`SortedSet<>` discards duplicates - if 2 requests happen to get the exact same expiry time then one of them would be lost and would never time out.  Strangely, `SortedList<>` also does not allow duplicates (it throws instead). The string comparison would be hit pretty rarely, I'd argue it's safer to leave it in.